skrifa/src/outline/autohint/latin.rs

```diff
@@ -0,0 +1,513 @@
+use raw::types::F2Dot14;
+use raw::TableProvider;
+
+impl UnscaledBlue {
+    fn is_latin_any_top(&self) -> bool {
+        self.flags & (blue_flags::LATIN_TOP | blue_flags::LATIN_SUB_TOP) != 0
+    }
+}
+
+impl UnscaledBlues {
+    /// Computes the set of blues for Latin style hinting.
+    ///
+    /// See <https://gitlab.freedesktop.org/freetype/freetype/-/blob/57617782464411201ce7bbc93b086c1b4d7d84a5/src/autofit/aflatin.c#L314>
+    pub fn new_latin(font: &FontRef, coords: &[F2Dot14], script: &ScriptClass) -> Self {
+        const MAX_INLINE_POINTS: usize = 64;
+        const BLUE_STRING_MAX_LEN: usize = 51;
+        let mut blues = Self::default();
+        let mut outline_buf = UnscaledOutlineBuf::<MAX_INLINE_POINTS>::new();
+        let mut flats = [0; BLUE_STRING_MAX_LEN];
+        let mut rounds = [0; BLUE_STRING_MAX_LEN];
+        let glyphs = font.outline_glyphs();
+        let charmap = font.charmap();
+        let units_per_em = font
+            .head()
+            .map(|head| head.units_per_em())
+            .unwrap_or_default() as i32;
+        let flat_threshold = units_per_em / 14;
+        // Walk over each of the blue character sets for our script.
+        for (blue_chars, blue_flags) in script.blues {
+            let is_top_like =
+                (blue_flags & (blue_flags::LATIN_TOP | blue_flags::LATIN_SUB_TOP)) != 0;
+            let is_top = blue_flags & blue_flags::LATIN_TOP != 0;
+            let is_x_height = blue_flags & blue_flags::LATIN_X_HEIGHT != 0;
+            let is_neutral = blue_flags & blue_flags::LATIN_NEUTRAL != 0;
+            let is_long = blue_flags & blue_flags::LATIN_LONG != 0;
+            let mut ascender = i16::MIN;
+            let mut descender = i16::MAX;
+            let mut n_flats = 0;
+            let mut n_rounds = 0;
+            for ch in *blue_chars {
+                // TODO: do some shaping
+                let y_offset = 0;
+                let Some(gid) = charmap.map(*ch) else {
+                    continue;
+                };
+                if gid.to_u32() == 0 {
+                    continue;
+                }
+                let Some(glyph) = glyphs.get(gid) else {
+                    continue;
+                };
+                outline_buf.clear();
+                if glyph.draw_unscaled(coords, None, &mut outline_buf).is_err() {
+                    continue;
+                }
+                let outline = outline_buf.as_ref();
+                // Reject glyphs that don't produce any rendering
+                if outline.points.len() <= 2 {
+                    continue;
+                }
+                let mut best_y: Option<i16> = None;
+                let mut best_y_extremum = if is_top { i32::MIN } else { i32::MAX };
+                let mut best_is_round = false;
+                // Find the extreme point depending on whether this is a top or
+                // bottom blue
+                let best_contour_and_point = if is_top_like {
```

Comment:
do you think there's any room to move any of this stuff into some little helpers or something? The diff gets very hard to read with all the right-drift...